### PR TITLE
Support maximisation for ecdf_logger

### PR DESCRIPTION
**Pull request by FurongYe**
By default, we think IOH_optimization_type::Miximization as using PBO_suite, and IOH_optimization_type::Minimization as using BBO_suite.

### CODE REVIEW

File: src/Template/Loggers/IOHprofiler_ecdf_logger.hpp [hunk 1]
```diff
@@ -251,23 +256,43 @@ void IOHprofiler_ecdf_logger<T>::fill_up( size_t i_error, size_t j_evals)
...
+                    if(mat[i-1][j] == 1) {
+                        jmax = j;
+                        break;
+                    } else {
+                        mat[i-1][j] = 1;
+                    }
+                } // for j
+            }
+        } // for i
+    }
```
Comment (wangronin) [inline src/Template/Loggers/IOHprofiler_ecdf_logger.hpp:295]:
```suggestion
    if(_current.maxmin == IOH_optimization_type::Minimization) {
        std::vector<size_t> v(imax - i_error - 1);
        std::generate(v.begin(), v.end(), [n = i_error] () mutable { return n++; });
    } else {
        std::vector<size_t> v(i_error + 1);
        std::generate(v.begin(), v.end(), [n = i_error] () mutable { return n--; });
    }

    for(auto i : v) {
        // If we reach a 1 on first col of this row, no need to continue.
        if(mat[i][j_evals] == 1) {
            break;
        } else {
            for(size_t j = j_evals; j < jmax; j++) {
                // If we reach a 1 on this col, no need to fill
                // the remaining columns for the next rows..
                if(mat[i][j] == 1) {
                    jmax = j;
                    break;
                } else {
                    mat[i][j] = 1;
                }
            } // for j
        }
    } // for i
```

**Comment by nojhan** [comment]
I see two fixes in this PR, one on the error computation and one the cuts during histogram construction.

I think there is some misunderstanding in the first fix. It tries to avoid relying on the known optimum by involving the optimization type (min/max). I can see that this fixes the bug, because PBO may not have a known optimum, whether BBOB have one. However, this will produce big problems on the long term. The correct fix would be to check for the existence of a known optimum, the error can then be use for both min or max problem without any drawback.

By the way, relying on direct access to the optimum vector is error prone, as it may return random (large) values if not initiliazed, without warning. I suggest to add accessors to this kind of members, so as to allow for checking if its value have been initialized (like, adding a flag `bool has_optimum` and asserts in the accessors, at least in Debug builds).

The second fix seems good, I would suggest to rename "*max" variables by "*bound", for the sake of generality.

File: src/Template/Loggers/IOHprofiler_ecdf_logger.hpp [hunk 2]
```diff
@@ -146,7 +146,12 @@ void IOHprofiler_ecdf_logger<T>::do_log(const std::vector<double>& infos)
 
     // TODO make a struct for loggerInfo?
     double evals = infos[0];
-    double err = std::abs(_current.opt[0] - infos[4]);
+    double err;
+    if (_current.maxmin == IOH_optimization_type::Minimization) {
+        err = std::abs(_current.opt[0] - infos[4]);
```
[the remark this answers is not in the document]
Comment (FurongYe) [reply src/Template/Loggers/IOHprofiler_ecdf_logger.hpp:151]: This is the target value that is used for BBOB problem.

File: src/Template/Loggers/IOHprofiler_ecdf_logger.hpp [hunk 3]
```diff
@@ -251,23 +256,43 @@ void IOHprofiler_ecdf_logger<T>::fill_up( size_t i_error, size_t j_evals)
...
+        for(size_t i = i_error; i < imax; i++) {
+            // If we reach a 1 on first col of this row, no need to continue.
+            if(mat[i][j_evals] == 1) {
+                break;
+            } else {
+                for(size_t j = j_evals; j < jmax; j++) {
+                    // If we reach a 1 on this col, no need to fill
+                    // the remaining columns for the next rows..
+                    if(mat[i][j] == 1) {
+                        jmax = j;
```
[the remark this answers is not in the document]
Comment (FurongYe) [reply src/Template/Loggers/IOHprofiler_ecdf_logger.hpp:269]: Because if it is 1, values on larger evaluations will be 1 as well.

**Comment by FurongYe** [comment]
> By the way, relying on direct access to the optimum vector is error prone, as it may return random (large) values if not initiliazed, without warning. I suggest to add accessors to this kind of members, so as to allow for checking if its value have been initialized (like, adding a flag `bool has_optimum` and asserts in the accessors, at least in Debug builds).

Thanks nojhan! Because for maximisation (PBO), we are looking fitness values, as you said, optimum may be unknown. I agree that we should set a flag for checking if the existence of optimum. Also, we will consider how/if we can set a consistent target variable for both minimisation and maximisation.

**Pull request merged**